perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java

```diff
@@ -3415,16 +3418,54 @@ public void deleteAllAttributeAuthz(PerunSession sess, AttributeDefinition attri
 		}
 
 		try {
-			return jdbc.query("SELECT " + getAttributeMappingSelectQuery("mem") + ", members.id FROM attr_names " +
-							"JOIN service_required_attrs ON attr_names.id=service_required_attrs.attr_id AND service_required_attrs.service_id=? " +
-							"JOIN members ON " + BeansUtils.prepareInSQLClause("members.id", members) +
-							"LEFT JOIN member_resource_attr_values mem ON attr_names.id=mem.attr_id AND mem.resource_id=? " +
-							"AND mem.member_id=members.id WHERE namespace IN (?,?,?)",
-					new MemberAttributeExtractor(sess, this, resource, members), service.getId(), resource.getId(),
-					AttributesManager.NS_MEMBER_RESOURCE_ATTR_DEF, AttributesManager.NS_MEMBER_RESOURCE_ATTR_OPT, AttributesManager.NS_MEMBER_RESOURCE_ATTR_VIRT);
-		} catch (RuntimeException ex) {
-			throw new InternalErrorException(ex);
-		}
+			if (Compatibility.isOracle()) {
+				return jdbc.execute("SELECT " + getAttributeMappingSelectQuery("mem") + ", members.id FROM attr_names " +
+					"JOIN service_required_attrs ON attr_names.id=service_required_attrs.attr_id AND service_required_attrs.service_id=? " +
+					"JOIN members ON members.id in (SELECT * FROM TABLE(?)) " +
+					"LEFT JOIN member_resource_attr_values mem ON attr_names.id=mem.attr_id AND mem.resource_id=? " +
+					"AND mem.member_id=members.id WHERE namespace IN (?,?,?)", (PreparedStatementCallback<HashMap<Member, List<Attribute>>>) preparedStatement -> {
+						Array sqlArray = ((PerunBlImpl)perun).getDatabaseManagerBl().prepareOracleArrayOfNumbers(members, preparedStatement);
+						preparedStatement.setInt(1, service.getId());
+						preparedStatement.setArray(2, sqlArray);
+						preparedStatement.setInt(3, resource.getId());
+						preparedStatement.setNString(4, AttributesManager.NS_MEMBER_RESOURCE_ATTR_DEF);
+						preparedStatement.setNString(5, AttributesManager.NS_MEMBER_RESOURCE_ATTR_OPT);
+						preparedStatement.setNString(6, AttributesManager.NS_MEMBER_RESOURCE_ATTR_VIRT);
+						MemberAttributeExtractor memberAttributeExtractor = new MemberAttributeExtractor(sess, this, resource, members);
+						return memberAttributeExtractor.extractData(preparedStatement.executeQuery());
+					});
+			}
+		} catch (InternalErrorRuntimeException ex) {
+			//This is possible if oracle sql array can't be created successfully (for example if oracle driver is missing)
+			//If happen, log it as error and continue using the standard way to obtain objects by IN clause
+			log.error("Unable to use Oracle SQL Array to obtain objects from database! Standard way to obtain these objects will be used instead.", ex);
+		}
+
+		HashMap<Member, List<Attribute>> allReturnedAttrs = new HashMap<>();
+		int from = 0;
+		int to = MAX_SIZE_OF_BULK_IN_SQL;
+
+		//Do not call one query for more than MAX_SIZE_OF_BULK_IN_SQL values in "SQL IN CLAUSE"
```

Comment:
The same trick as for Oracle can be done also for PostgreSQL, just use 
`WHERE id = ANY(?)` instead of `WHERE id IN (SELECT * FROM TABLE(?))` and use 
```
Array userIds = con.createArrayOf(JDBCType.INTEGER.name(), new Integer[] { 1,2,3 });
```
to create the array.

Reply:
The same can be done for HSQL, just use 
`WHERE id IN ( UNNEST(?) )` instead of `WHERE id = ANY(?)`